LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/Utils/TwinCollectionExtensions.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+namespace LoRaTools.Utils
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Globalization;
+    using LoRaWan;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    public static class TwinCollectionExtensions
+    {
+        private static readonly Dictionary<Type, Func<object, object>> customConverters = new Dictionary<Type, Func<object, object>>
+        {
+            [typeof(StationEui)] = (s) => StationEui.Parse(s.ToString()),
+            [typeof(DevNonce)] = (s) => new DevNonce(Convert.ToUInt16(s, CultureInfo.InvariantCulture)),
+            [typeof(DevAddr)] = (s) => DevAddr.Parse(s.ToString()),
+        };
+
+        public static T? SafeRead<T>(this TwinCollection twinCollection, string property, T? defaultValue = default, ILogger? logger = null)
+            => twinCollection.TryRead<T>(property, logger, out var someT) ? someT : defaultValue;
+
+        public static string ReadRequiredString(this TwinCollection twinCollection, string property, ILogger? logger = null) =>
+                twinCollection.TryRead<string>(property, logger, out var someString) && !string.IsNullOrEmpty(someString)
+                                ? someString
+                                : throw new InvalidOperationException($"Property '{property}' does not exist or is empty.");
+
+        public static bool TryRead<T>(this TwinCollection twinCollection, string property, ILogger? logger, [NotNullWhen(true)] out T? value)
+        {
+            _ = twinCollection ?? throw new ArgumentNullException(nameof(twinCollection));
+
+            value = default;
+
+            if (!twinCollection.Contains(property))
+                return false;
+
+            // cast to object to avoid dynamic code to be generated
+            var some = (object)twinCollection[property];
+
+            // quick path for values that can be directly converted
+            if (some is JValue someJValue)
+            {
+                if (someJValue.Value is T someT)
+                {
+                    value = someT;
+                    return true;
+                }
+
+                if (someJValue.Type == JTokenType.Null)
+                {
+                    return false;
+                }
+            }
+
+            try
+            {
+                var t = typeof(T);
+
+                if (TryGetCustomConverter(t, out var converter))
```

Comment:
`TryGetCustomConverter` has only one reference and could be inline to avoid indirection. Alternatively, we could also get rid of the `customConverters` entirely and use something similar to:

```csharp
value = value switch
{
    StationEui => (T)(object)StationEui.Parse(some.ToString()),
    DevNonce => (T)(object)new DevNonce(Convert.ToUInt16(some, CultureInfo.InvariantCulture)),
    DevAddr => (T)(object)DevAddr.Parse(some.ToString()),
    _ => (T)Convert.ChangeType(some, t, CultureInfo.InvariantCulture)
};

if (t.IsEnum && !t.IsEnumDefined(value))
{
    value = default;
    LogParsingError(logger, property, some);
    return false;
}
```

Reply:
I think the case of custom converters is different and we do not want to go through the additional code we are having in there (enum validation). 

The reason for the additional method is the nullable types. The dictionary can't indicate that the out is not null when it returns true, hence the redirection. I'll keep it the way it is.